src/plonk/circuit.rs

```diff
@@ -775,10 +771,72 @@ impl<F: Field> ConstraintSystem<F> {
 
         // Account for each gate to ensure our quotient polynomial is the
         // correct degree and that our extended domain is the right size.
-        for (_, poly) in self.gates.iter() {
-            degree = std::cmp::max(degree, poly.degree());
+        for gate in &self.gates {
+            degree = std::cmp::max(degree, gate.poly().degree());
         }
 
         degree
     }
 }
+
+/// Exposes the "virtual registers" that can be queried while creating a custom gate or
+/// lookup table.
+#[derive(Debug)]
+pub struct Registers<'a, F: Field> {
+    meta: &'a mut ConstraintSystem<F>,
+    queried_selectors: Vec<Selector>,
+    queried_columns: Vec<Column<Any>>,
+}
+
+impl<'a, F: Field> Registers<'a, F> {
+    fn new(meta: &'a mut ConstraintSystem<F>) -> Self {
+        Registers {
+            meta,
+            queried_selectors: vec![],
+            queried_columns: vec![],
+        }
+    }
+
+    /// Query a selector at a relative position.
+    pub fn query_selector(&mut self, selector: Selector, at: Rotation) -> Expression<F> {
+        self.queried_selectors.push(selector);
+        Expression::Fixed(self.meta.query_fixed_index(selector.0, at))
+    }
+
+    /// Query a fixed column at a relative position
+    pub fn query_fixed(&mut self, column: Column<Fixed>, at: Rotation) -> Expression<F> {
+        self.queried_columns.push(column.into());
```

Comment:
Currently this only tracks the queried columns, but I'm not sure if that's sufficient for the kinds of debug checks we want to perform. I think we might actually want to track queried "virtual registers", being the tuple `(Column<Any>, Rotation)`.

Reply:
Mmm, now that I think about it more, I think this is indeed what we'd want to track. Then when a selector is enabled on a row, we can figure out all of the relative cells that need to be assigned for this gate to be useful.